tools/yul_phaser/Chromosome.cpp

```diff
@@ -0,0 +1,190 @@
+#include <libyul/optimiser/RedundantAssignEliminator.h>
+#include <libyul/optimiser/Rematerialiser.h>
+#include <libyul/optimiser/SSAReverser.h>
+#include <libyul/optimiser/SSATransform.h>
+#include <libyul/optimiser/StructuralSimplifier.h>
+#include <libyul/optimiser/Suite.h>
+#include <libyul/optimiser/UnusedPruner.h>
+#include <libyul/optimiser/VarDeclInitializer.h>
+#include <libyul/optimiser/VarNameCleaner.h>
+
+#include <cassert>
+#include <sstream>
+
+using namespace std;
+using namespace solidity::yul;
+
+namespace solidity::phaser
+{
+
+Chromosome Chromosome::make_random(int _length)
+{
+	assert(_length >= 0);
+
+	vector<string> steps;
+	steps.reserve(_length);
+	for (auto i = 0; i < _length; ++i)
+		steps.push_back(randomOptimizationStep());
+
+	return Chromosome(move(steps));
+}
+
+ostream& operator<<(ostream& _stream, Chromosome const& _chromosome)
+{
+	for (auto stepName: _chromosome.m_optimizationSteps)
+		_stream << Chromosome::abbreviateStepName(stepName);
+
+	return _stream;
+}
+
+char Chromosome::abbreviateStepName(string const& _stepName)
+{
+	static unordered_map<string, char> stepToSymbol{
+		{BlockFlattener::name,                'f'},
+		{CommonSubexpressionEliminator::name, 'c'},
+		{ConditionalSimplifier::name,         'C'},
+		{ConditionalUnsimplifier::name,       'U'},
+		{ControlFlowSimplifier::name,         'n'},
+		{DeadCodeEliminator::name,            'D'},
+		{EquivalentFunctionCombiner::name,    'v'},
+		{ExpressionInliner::name,             'e'},
+		{ExpressionJoiner::name,              'j'},
+		{ExpressionSimplifier::name,          's'},
+		{ExpressionSplitter::name,            'x'},
+		{ForLoopConditionIntoBody::name,      'I'},
+		{ForLoopConditionOutOfBody::name,     'O'},
+		{ForLoopInitRewriter::name,           'o'},
+		{FullInliner::name,                   'i'},
+		{FunctionGrouper::name,               'g'},
+		{FunctionHoister::name,               'h'},
+		{LiteralRematerialiser::name,         'T'},
+		{LoadResolver::name,                  'L'},
+		{LoopInvariantCodeMotion::name,       'M'},
+		{RedundantAssignEliminator::name,     'r'},
+		{Rematerialiser::name,                'm'},
+		{SSAReverser::name,                   'V'},
+		{SSATransform::name,                  'a'},
+		{StructuralSimplifier::name,          't'},
+		{UnusedPruner::name,                  'u'},
+		{VarDeclInitializer::name,            'd'},
+		{VarNameCleaner::name,                'l'},
+	};
+
+	auto pair = stepToSymbol.find(_stepName);
+	assert(pair != stepToSymbol.end());
+
+	return pair->second;
+}
+
+vector<string> Chromosome::allStepNames()
+{
+	assert(not OptimiserSuite::allSteps().empty());
+
+	vector<string> stepNames;
+	stepNames.reserve(OptimiserSuite::allSteps().size());
+
+	for (auto const& step: OptimiserSuite::allSteps())
+		stepNames.push_back(step.first);
+
+	return stepNames;
+}
+
+vector<string> Chromosome::allStepNamesExcept(vector<string> const& _excludedStepNames)
+{
+	vector<string> all = allStepNames();
```

Comment:
Maybe - half jokingly:
```
return convert_container<vector<string>>(convert_container<set<string>>(allStepNames()) - convert_container<set<string>>(_excludedStepNames));
```

Reply:
I wish! Using set arithmetic is actually the first thing that comes to my mind in a situation like this but looks like `std::set` does not implement the `-` operator :( There's `set_difference()` but it's not as concise so I settled for a boring `for` :)

Reply:
We have an overload for `operator-` in one of the util headers.

Reply:
Done.

Actually I removed this function a commit later anyway because after removing `VarNameCleaner` nothing needs to be excluded any more, but it's still in commit history in case I need it back. I bet it'll come in handy for debugging later.